target/linux/ipq40xx/base-files/lib/upgrade/linksys.sh

```diff
@@ -0,0 +1,101 @@
+#
+# Copyright (C) 2014-2015 OpenWrt.org
+#
+
+linksys_get_target_firmware() {
+	cur_boot_part=`/usr/sbin/fw_printenv -n boot_part`
+	target_firmware=""
+	if [ "$cur_boot_part" = "1" ]
+	then
+		# current primary boot - update alt boot
+		target_firmware="alt_kernel"
+		fw_setenv boot_part 2
+		#In EA6350v3 this causes the device to reboot in the old partition
+		#fw_setenv bootcmd "run altnandboot"
+	elif [ "$cur_boot_part" = "2" ]
+	then
+		# current alt boot - update primary boot
+		target_firmware="kernel"
+		fw_setenv boot_part 1
+		#In EA6350v3 this causes the device to reboot in the old partition
+		#fw_setenv bootcmd "run nandboot"
+	fi
+
+	# re-enable recovery so we get back if the new firmware is broken
+	fw_setenv auto_recovery yes
+
+	echo "$target_firmware"
+}
+
+linksys_get_root_magic() {
+	(get_image "$@" | dd skip=786432 bs=4 count=1 | hexdump -v -n 4 -e '1/1 "%02x"') 2>/dev/null
+}
+
+platform_do_upgrade_linksys() {
+	local magic_long="$(get_magic_long "$1")"
+
+	mkdir -p /var/lock
+	local part_label="$(linksys_get_target_firmware)"
+	touch /var/lock/fw_printenv.lock
+
+	if [ ! -n "$part_label" ]
+	then
+		echo "cannot find target partition"
+		exit 1
+	fi
+
+	local target_mtd=$(find_mtd_part $part_label)
+
+	[ "$magic_long" = "73797375" ] && {
+		CI_KERNPART="$part_label"
+		if [ "$part_label" = "kernel" ]
+		then
+			CI_UBIPART="rootfs"
+		else
+			CI_UBIPART="alt_rootfs"
+		fi
+
+
+		# remove "squashfs" vol (in case we are flashing over a stock image, which is also UBI)
+
+		local mtdnum="$( find_mtd_index "$CI_UBIPART" )"
+		if [ ! "$mtdnum" ]; then
+			echo "cannot find ubi mtd partition $CI_UBIPART"
+			return 1
+		fi
+
+		local ubidev="$( nand_find_ubi "$CI_UBIPART" )"
+		if [ ! "$ubidev" ]; then
+			ubiattach -m "$mtdnum"
+			sync
+			ubidev="$( nand_find_ubi "$CI_UBIPART" )"
+		fi
+
+		if [ "$ubidev" ]; then
+
+			local squash_ubivol="$( nand_find_volume $ubidev squashfs )"
+
+			# kill volume
+			[ "$squash_ubivol" ] && ubirmvol /dev/$ubidev -N squashfs || true
+		fi
+
+
+		# complete std upgrade
+		nand_upgrade_tar "$1"
+	}
+	[ "$magic_long" = "27051956" ] && {
+		# check firmwares' rootfs types
+		local oldroot="$(linksys_get_root_magic $target_mtd)"
+		local newroot="$(linksys_get_root_magic "$1")"
+
+		if [ "$newroot" = "55424923" -a "$oldroot" = "55424923" ]
+		# we're upgrading from a firmware with UBI to one with UBI
+		then
+			# erase everything to be safe
+			mtd erase $part_label
```

Comment:
It's a bit odd that the mtd gets erased by a extra command.
and the following mtd write gets the -n (write without first erasing the blocks) flag.

Is this code-path used to revert back to stock, or what is going on?

Reply:
It is in fact not "needed", but old software from the other partition will be present as the /overlay (rootfs_data) was not erased.
This is not that bad for images with the same number of blocks but if you flash either bigger or smaller partitions, the /overlay partition (rootfs_data) will be corrupted anyway.

Reply:
> Forget this... I don't know what happend. Sometimes the device boots to the old partition but I don't know why.

I think it should be possible to do this all in a simpler way. You have to make sure that rootfs and alt_rootfs are mapped to mtd11 and mtd13 respectively and then you could just do something like the following example (untested):

```
platform_do_upgrade_linksys() {
      CI_KERNPART="$(linksys_get_target_firmware)"
      case "$CI_KERNPART" in
      kernel)
          CI_UBIPART="rootfs"
          ;;
      alt_kernel)
          CI_UBIPART="alt_rootfs"
          ;;
      *)
          # something is wrong, very wrong.
          echo "Help error message about uboot-env being messed up and that the device is probably going to be bricked. But there's nothing that can be done because this script will reboot either way :("
          exit 1
       ;;
       esac
       nand_do_upgrade "$ARGV"
}
```

(Note: I don't know how the "back-to-stock" flash is supposed to be done. If you think you'll can't live without it then please add it)